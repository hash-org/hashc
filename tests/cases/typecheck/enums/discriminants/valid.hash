// run=pass, stage=typecheck

#layout_of
#[repr("i8")]
Direction := enum(
    #[discriminant(-120)]
    Up,
    Right,
    Down,
    Left,
)

#layout_of
ShipKind := enum(
    #[discriminant(-120)]
    Cruise,
    Jet,
    Steamer,
    SpeedBoat,
)

#layout_of
ModuleCode := enum(
    #[discriminant(3050)]
    Logic,
    #[discriminant(3052)]
    Complexity,
    #[discriminant(4201)]
    PLDI,
)


FloorId := struct(i8);

#layout_of
EmployeeKind := enum(
    #[discriminant(7)]
    FactoryWorker(FloorId),
    #[discriminant(12)]
    Engineer,
    #[discriminant(83)]
    Hardware,
    #[discriminant(666)]
    Manager,
)




main := () => {}
