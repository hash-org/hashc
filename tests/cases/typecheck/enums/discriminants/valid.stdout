info: Layout of `Direction` (size=1b align=1b):
┌─────────────┐
│ 1b tag (i8) │
├─────────────┤
│     Up      │
│             │
│    -120     │
├─────────────┤
│    Right    │
│             │
│    -119     │
├─────────────┤
│    Down     │
│             │
│    -118     │
├─────────────┤
│    Left     │
│             │
│    -117     │
└─────────────┘

info: Layout of `ShipKind` (size=2b align=2b):
┌──────────────┐
│ 2b tag (i16) │
├──────────────┤
│    Cruise    │
│              │
│     -120     │
├──────────────┤
│     Jet      │
│              │
│     -119     │
├──────────────┤
│   Steamer    │
│              │
│     -118     │
├──────────────┤
│  SpeedBoat   │
│              │
│     -117     │
└──────────────┘

info: Layout of `ModuleCode` (size=2b align=2b):
┌──────────────┐
│ 2b tag (u16) │
├──────────────┤
│    Logic     │
│              │
│     3050     │
├──────────────┤
│  Complexity  │
│              │
│     3052     │
├──────────────┤
│     PLDI     │
│              │
│     4201     │
└──────────────┘

info: Layout of `EmployeeKind` (size=4b align=2b):
┌───────────────┬─────────────────────┐
│ 2b tag (u16)  │                     │
├───────────────┼────────────┬────────┤
│ FactoryWorker │ 0: FloorId │ 1b pad │
│               │            │        │
│       7       │   size: 1b │   ##   │
│               │ offset: 2b │   ##   │
│               │  align: 2b │   ##   │
├───────────────┼────────────┴────────┤
│   Engineer    │                     │
│               │                     │
│      12       │                     │
├───────────────┼─────────────────────┤
│   Hardware    │                     │
│               │                     │
│      83       │                     │
├───────────────┼─────────────────────┤
│    Manager    │                     │
│               │                     │
│      666      │                     │
└───────────────┴─────────────────────┘

info: Layout of `Variants` (size=2b align=2b):
┌──────────────┐
│ 2b tag (u16) │
├──────────────┤
│      A       │
│              │
│     255      │
├──────────────┤
│      B       │
│              │
│     256      │
└──────────────┘

info: Layout of `!` (size=0b align=1b):
┌────────────┐
│  <ZST>: !  │
│            │
│   size: 0b │
│ offset: 0b │
│  align: 1b │
└────────────┘

info: Layout of `Boon` (size=16b align=8b):
┌──────────────┬─────────────────────────┐
│ 4b tag (u32) │                         │
├──────────────┼────────────┬────────────┤
│     Foo      │   x: i32   │   y: i32   │
│              │            │            │
│      0       │   size: 4b │   size: 4b │
│              │ offset: 4b │ offset: 8b │
│              │  align: 4b │  align: 4b │
├──────────────┼────────┬───┴────────────┤
│     Faz      │ 4b pad │     z: f64     │
│              │        │                │
│      1       │   ##   │     size: 8b   │
│              │   ##   │   offset: 8b   │
│              │   ##   │    align: 8b   │
└──────────────┴────────┴────────────────┘

